README.md
# Video Analyst

[![Build Status](https://travis-ci.org/MegviiDetection/video_analyst.svg?branch=master)](https://travis-ci.org/MegviiDetection/video_analyst)

This is the implementation of a series of basic algorithms which is useful for video understanding, including Single Object Tracking (SOT), Video Object Segmentation (VOS), etc.

Current implementation list:

* SOT
  * [SiamFC++: Towards Robust and Accurate Visual Tracking with Target Estimation Guidelines](https://arxiv.org/abs/1911.06188) [[demo]](https://www.youtube.com/watch?v=TCziWahnXT8&list=PL4KqNq8e6fJkfk35zHRaUd21ExV522JK0&index=4&t=0s&app=desktop)
<div align="center">
  <img src="docs/resources/siamfcpp_ice2.gif" width="800px" />
  <p>Example SiamFC++ outputs.</p>
</div>

* VOS
  * [State-Aware Tracker for Real-Time Video Object Segmentation](https://arxiv.org/abs/2003.00482) 
<div align="center">
  <img src="docs/resources/sat_runman.gif" width="800px" />
  <p>Example SAT outputs.</p>
</div>

## SOT Quick start

### Setup

Please refer to [SOT_SETUP.md](docs/TUTORIALS/SOT_SETUP.md)
### Demo

#### SOT video demo
```Bash
# demo with web camera
python3 ./demo/main/video/sot_video.py --config 'experiments/siamfcpp/test/vot/siamfcpp_alexnet.yaml' --device cuda --video "webcam" 

# demo with video file, and dump result into video file (optional)
python3 ./demo/main/video/sot_video.py --config 'experiments/siamfcpp/test/vot/siamfcpp_alexnet.yaml' --device cuda --video $video_dir/demo.mp4 --output $dump_path/result.mp4

# demo with extracted image files, and dump result into image files (optional)
python3 ./demo/main/video/sot_video.py --config 'experiments/siamfcpp/test/vot/siamfcpp_alexnet.yaml' --device cuda --video $video_dir/*.jpg --output $dump_dir
```

### Test

Please refer to [SOT_TEST.md](docs/TUTORIALS/SOT_TEST.md) for detail.

### Training

Please refer to [SOT_TRAINING.md](docs/TUTORIALS/SOT_TRAINING.md) for detail.

## Repository structure (in progress)

```File Tree
project_root/
├── experiments  # experiment configurations, in yaml format
├── main
│   ├── train.py  # trainng entry point
│   └── test.py  # test entry point
├── video_analyst
│   ├── data  # modules related to data
│   │   ├── dataset  # data fetcher of each individual dataset
│   │   ├── sampler  # data sampler, including inner-dataset and intra-dataset sampling procedure
│   │   ├── dataloader.py  # data loading procedure
│   │   └── transformer  # data augmentation
│   ├── engine  # procedure controller, including traiing control / hp&model loading
│   │   ├── monitor  # monitor for tasks during training, including visualization / logging / benchmarking
│   │   ├── trainer.py  # train a epoch
│   │   ├── tester.py  # test a model on a benchmark
│   ├── model # model builder
│   │   ├── backbone  # backbone network builder
│   │   ├── common_opr  # shared operator (e.g. cross-correlation)
│   │   ├── task_model  # holistic model builder
│   │   ├── task_head  # head network builder
│   │   └── loss  # loss builder
│   ├── pipeline  # pipeline builder (tracking / vos)
│   │   ├── segmenter  # segmenter builder for vos
│   │   ├── tracker  # tracker builder for tracking
│   │   └── utils  # pipeline utils
│   ├── config  # configuration manager
│   ├── evaluation  # benchmark
│   ├── optim  # optimization-related module (learning rate, gradient clipping, etc.)
│   │   ├── optimizer # optimizer
│   │   ├── scheduler # learning rate scheduler
│   │   └── grad_modifier # gradient-related operation (parameter freezing)
│   └── utils  # useful tools
└── README.md
```

## docs

For detail, please refer to markdown files under _docs_.

### SOT 
* [SOT_SETUP.md](docs/TUTORIALS/SOT_SETUP.md): instructions for setting-up
* [SOT_MODEL_ZOO.md](docs/TUTORIALS/SOT_MODEL_ZOO.md): description of released sot models
* [SOT_TRAINING.md](docs/TUTORIALS/SOT_TRAINING.md): details related to training
* [SOT_TEST.md](docs/TUTORIALS/SOT_TEST.md): details related to test
### VOS 
* [VOS_SETUP.md](docs/TUTORIALS/VOS_SETUP.md): instructions for setting-up
* [VOS_MODEL_ZOO.md](docs/TUTORIALS/VOS_MODEL_ZOO.md): description of released sot models
* [VOS_TRAINING.md](docs/TUTORIALS/VOS_TRAINING.md): details related to training
* [VOS_TEST.md](docs/TUTORIALS/VOS_TEST.md): details related to training
### DEVELOP
* [DEVELOP.md](docs/DEVELOP/DEVELOP.md): description of project design (registry, configuration tree, etc.)
* [PIPELINE_API.md](docs/DEVELOP/PIPELINE_API.md): description for pipeline API
* [FORMATTING_INSTRUCTION](docs/DEVELOP/REFORMATTING_INSTRUCTIONS.md): instruction for code formatting (yapf/isort/flake/etc.)

## TODO
[] refine code stype and test cases

## Acknowledgement

* video_analyst/evaluation/vot_benchmark and other related code have been borrowed from [PySOT](https://github.com/STVIR/pysot)
* video_analyst/evaluation/got_benchmark and other related code have been borrowed from [got-toolkit](https://github.com/got-10k/toolkit.git)
* [detectron2](https://github.com/facebookresearch/detectron2)
* [fvcore](https://github.com/facebookresearch/fvcore)
* [pytracking](https://github.com/visionml/pytracking/tree/f0fc646b98166cc28c33f6b12f5169bb6f0647be)
* [DROL](https://github.com/shallowtoil/DROL)

## References

```
@inproceedings{xu2020siamfc++,
  title={SiamFC++: Towards Robust and Accurate Visual Tracking with Target Estimation Guidelines.},
  author={Xu, Yinda and Wang, Zeyu and Li, Zuoxin and Yuan, Ye and Yu, Gang},
  booktitle={AAAI},
  pages={12549--12556},
  year={2020}
}
```
```
@inproceedings{chen2020state,
  title={State-Aware Tracker for Real-Time Video Object Segmentation},
  author={Chen, Xi and Li, Zuoxin and Yuan, Ye and Yu, Gang and Shen, Jianxin and Qi, Donglian},
  booktitle={Proceedings of the IEEE/CVF Conference on Computer Vision and Pattern Recognition},
  pages={9384--9393},
  year={2020}
}
```
## Contact

Maintainer (sorted by family name):

* Xi Chen[@XavierCHEN34](https://github.com/XavierCHEN34)
* Zuoxin Li[@lzx1413](https://github.com/lzx1413)
* Zeyu Wang[@JWarlock](http://github.com/JWarlock)
* Yinda Xu[@MARMOTatZJU](https://github.com/MARMOTatZJU)


CHANGELOG.md
# Changelog

Date format: YY/MM/DD

## 2020/08/15
* add atom online module from pytracking & DROL
* update speed defination in vot test

## 2020/08/08
* adopt pre-commit to format code while committing
* reformat code with isort, autoflake and yapf

## 2020/07/30
* support amp from pytorch>=1.6

## 2020/06/29
* support acclerating tracking by tensorRT

## 2020/06/23

* refine make_densebox_target.py
  * target making code: numpy -> pytorch
  * simplified for SOT task
    * reduce to single resolution layer
* relax restrictions in requirements.txt
  * e.g. version of pytorch, opencv, numpy, etc.

## 2020/05/08
* format evaluation code
* Long-term tracking 
   * add dataset and test_engine for UAV123 and VOTLT
   * add test results for UAV123
   * add long-term trakcker (need to  be refined)
   
## 2020/04/28

* Modify interface of _siamfcpp_track.py: update_
  * Accept provided target state (bbox)

## 2020/04/25

* Add multi-processing tester
  * otb
  * lasot
  * trackingnet
  * got10k
  
## 2020/04/18
* add SAT train code
* update SAT test code
* change TrackInfo to TextInfo
* new param loader for model borrowed from dectron2
* update coco dataset from detectron2

## 2020/04/18
* New features in SOT demo (sot_video.py)
  * Add image file video reader (reading from image files)
  * Add image file video dumping (dump result as frame-level image files)

## 2020/04/13
* Add OTB test

## 2020/04/12
* Add iterable dataset in
  * dataset_type under config node: data
    * regular: former dataset
    * iterable: newly added dataset
  * data/adaptor_dataset.py: _AdaptorIterableDataset_ / _MultiStreamDataloader_
  * data/builder.py: the branch with _cfg.dataset_type_
* Fix seed issue
  * pass builder's seed to dataset as ext_seed

## 2020/04/06
* Add model/pipeline/tester for SAT on DAVIS
* Add transferred models for SAT

## 2020/03/25

____Major changes____

* Add hyper-parameter search script
  * _main/hpo.py_
  * _experiment_
  * _videoanalyst/utils/hpo.py_

## 2020/03/23

____Major changes____

* Add full training data
  * coco / det / vid / trackingnet / lasot / got10k
  * with caching mechanism (reducing storage I/O by caching indexes into memory)
  * TODO: perform verification with full data (under batchsize=256)

## 2020/03/25

____Major changes____

* Add hyper-parameter search script
  * _main/hpo.py_
  * _experiment_
  * _videoanalyst/utils/hpo.py_

## 2020/03/23

____Major changes____

* Add full training data
  * coco / det / vid / trackingnet / lasot / got10k
  * with caching mechanism (reducing storage I/O by caching indexes into memory)
  * TODO: perform verification with full data (under batchsize=256)

## 2020/03/18

____Major changes____

* Add caching in LaSOT dataset
* Add multi-mode in siamfcpp visualization debugging tools
  * demo/main/debug/visualize_siamfcpp_training_data.py
  * Seperate config file for debugging: experiments/siamfcpp/data/siamfcpp_data-trn.yaml

## 2020/03/17
  
____Major changes____

* Update all training configuration
  * the cuda device number is defined by num_processes rather than devices in training phase
  * the settings between DP and DDP are unified now
* Adopt loguru to replace logging for better logger
* Update code for distributed training
* Update code for model/pipeline/trainer

## 2020/03/14

____Major changes____

* Update all training configuration
  * learning rate multiplied by 2 in order to be compatible with PyTorch==1.4.0 & CUDA==10.1
  * Update requirements.txt
* Add tinyconv training configuration
  * _experiments/siamfcpp/train/siamfcpp_tinyconv-trn.yaml_
* Add dataloader visualization tools for SiamFC++ (SOT)
  * _demo/main/debug/visualize_siamfcpp_training_data.py_

## 2020/03/07

____Major changes____

* Add webcam demo
  * _demo/main/video/sot_video.py_

## 2020/03/02

____Major changes____

* Add training with PyTorch Distributed Data Parallel (DDP)
  * _main/dist_trian.py_
  * _.../trainer_impl/distributed_regular_trainer.py_

## 2020/02/27

____Major changes____

* Add TensorboardLogger in __monitor__
  * Recursively put engine_data as scalar into Tensorboard

## 2020/02/22

____Major changes____

* Caching mechanism added to [got10k dataset](videoanalyst/evaluation/got_benchmark/datasets/got10k.py)
  * Build cache at _root_dir/subset.pkl_ (default), subset=[train, val, test]
  * Cache created/loaded in GOT10k.data_dict and will be queried every time _\_\_getitem\_\__ is called
* Remove _num_iterations_ item in training .yaml files
  * replaced by value calculated with _nr_image_per_epoch_ and _minibatch_
    * _num_iterations_ = _nr_image_per_epoch_ // _minibatch_
  * all training configs has been updated.

____Minor changes____

* pipeline.builder's builder method has been changed
  * _build_pipeline_ -> _build_

## 2020/02/15

* Add contrib module's template
  * _docs/TEMPLATES/contrib_module_

## 2020/02/14

* Add one-shot detection demo
  * API (see [docs/PIPELINE_API.md](docs/PIPELINE_API.md)
  * Runnable demo
* Complete ShuffleNetV2x1.0 experiment

## 2020/02/13

* Support Training and Test on GOT-10k for Single Object Tracking
  * SiamFC++-AlexNet
  * SiamFC++-GoogLeNet
  * SiamFC++-ShuffleNetV2
* Training Components
  * Dataset helper, data sampler, transformer (data augmentation), etc.
  * Optimization: learing rate scheduler, dynamic freezing, etc.
  * Trainer & Monitor

## 2019/12/31

* Support SiamFC++ test on VOT benchmark
  * SiamFC++-AlexNet
  * SiamFC++-GoogLeNet



main/cvt_onnx.py
# -*- coding: utf-8 -*-
from paths import ROOT_PATH  # isort:skip

import argparse
import os.path as osp

import numpy as np
import onnx
import onnxruntime
from loguru import logger

import torch
import torch.onnx

from videoanalyst.config.config import cfg as root_cfg
from videoanalyst.config.config import specify_task
from videoanalyst.model import builder as model_builder
from videoanalyst.utils import complete_path_wt_root_in_cfg


def make_parser():
    parser = argparse.ArgumentParser(description='Test')
    parser.add_argument('-cfg',
                        '--config',
                        default='',
                        type=str,
                        help='experiment configuration')
    parser.add_argument('-o',
                        '--output',
                        default='',
                        type=str,
                        help='output onnx file name')

    return parser


def to_numpy(tensor):
    return tensor.detach().cpu().numpy(
    ) if tensor.requires_grad else tensor.cpu().numpy()


def export_siamfcpp_fea_onnx(task_cfg, parsed_args):
    # build model
    model = model_builder.build("track", task_cfg.model)
    model.eval()
    # extract feature extractor onnx model
    batch_size = 1
    x = torch.randn(batch_size, 3, 127, 127)
    fea = model(x, phase="feature")
    output_path = parsed_args.output + "_fea.onnx"
    torch.onnx.export(
        model,
        x,
        output_path,
        verbose=False,
        export_params=True,
        opset_version=10,
        do_constant_folding=True,
        input_names=["target_img"],
        output_names=["c_z_k", "r_z_k"],
        dynamic_axes={
            'target_img': {
                0: 'batch_size'
            },  # variable lenght axes
            'c_z_k': {
                0: 'batch_size'
            },
            'r_z_k': {
                0: 'batch_size'
            }
        })
    # check onnx
    onnx_model = onnx.load(output_path)
    onnx.checker.check_model(onnx_model)
    ort_session = onnxruntime.InferenceSession(output_path)
    ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(x)}
    ort_outs = ort_session.run(None, ort_inputs)
    # compare ONNX Runtime and PyTorch results
    np.testing.assert_allclose(to_numpy(fea[0]),
                               ort_outs[0],
                               rtol=1e-03,
                               atol=1e-05)


def export_siamfcpp_track_onnx(task_cfg, parsed_args):
    # build model
    model = model_builder.build("track", task_cfg.model)
    model.eval()
    # extract feature extractor onnx model
    batch_size = 1
    search_im = torch.randn(batch_size, 3, 303, 303)
    c_z_k = torch.randn(batch_size, 256, 4, 4)
    r_z_k = torch.randn(batch_size, 256, 4, 4)
    fea = model(search_im, c_z_k, r_z_k, phase="onnx_track")
    output_path = parsed_args.output + "_track.onnx"
    torch.onnx.export(model, (search_im, c_z_k, r_z_k),
                      output_path,
                      verbose=False,
                      export_params=True,
                      opset_version=10,
                      do_constant_folding=True,
                      input_names=["search_img", "c_z_k", "r_z_k"],
                      output_names=["cls", "ctr", "box"])
    # check onnx
    onnx_model = onnx.load(output_path)
    onnx.checker.check_model(onnx_model)
    ort_session = onnxruntime.InferenceSession(output_path)
    ort_inputs = {
        ort_session.get_inputs()[0].name: to_numpy(search_im),
        ort_session.get_inputs()[1].name: to_numpy(c_z_k),
        ort_session.get_inputs()[2].name: to_numpy(r_z_k)
    }
    ort_outs = ort_session.run(None, ort_inputs)
    # compare ONNX Runtime and PyTorch results
    np.testing.assert_allclose(to_numpy(fea[0]),
                               ort_outs[0],
                               rtol=1e-03,
                               atol=1e-05)
    np.testing.assert_allclose(to_numpy(fea[1]),
                               ort_outs[1],
                               rtol=1e-03,
                               atol=1e-05)


if __name__ == '__main__':
    # parsing
    parser = make_parser()
    parsed_args = parser.parse_args()

    # experiment config
    exp_cfg_path = osp.realpath(parsed_args.config)
    root_cfg.merge_from_file(exp_cfg_path)
    logger.info("Load experiment configuration at: %s" % exp_cfg_path)

    # resolve config
    root_cfg = complete_path_wt_root_in_cfg(root_cfg, ROOT_PATH)
    root_cfg = root_cfg.test
    task, task_cfg = specify_task(root_cfg)
    task_cfg.freeze()
    export_siamfcpp_track_onnx(task_cfg, parsed_args)
